cpp/CMakeLists.txt

```diff
@@ -783,24 +543,133 @@ set_property(CACHE RMM_LOGGING_LEVEL PROPERTY STRINGS
+        "${CMAKE_SOURCE_DIR}/src/transform"
+        "${CMAKE_SOURCE_DIR}/src/unary"
+        "${CMAKE_SOURCE_DIR}/src/utilities"
+)
+
+set(
+    CUDF_MODULE_LIST
+        "${CMAKE_SOURCE_DIR}/src/ast"
+        "${CMAKE_SOURCE_DIR}/src/comms"
+        "${CMAKE_SOURCE_DIR}/src/hash"
+        "${CMAKE_SOURCE_DIR}/src/interop"
+        "${CMAKE_SOURCE_DIR}/src/io"
+        "${CMAKE_SOURCE_DIR}/src/join"
+        "${CMAKE_SOURCE_DIR}/src/merge"
+        "${CMAKE_SOURCE_DIR}/src/partitioning"
+        "${CMAKE_SOURCE_DIR}/src/reductions"
+        "${CMAKE_SOURCE_DIR}/src/replace"
+        "${CMAKE_SOURCE_DIR}/src/reshape"
+        "${CMAKE_SOURCE_DIR}/src/rolling"
+        "${CMAKE_SOURCE_DIR}/src/transpose"
+)
+
+foreach(CUDF_BASE_MODULE_DIR IN LISTS CUDF_BASE_MODULE_LIST)
+    file(GLOB_RECURSE CUDF_BASE_FILES "${CUDF_BASE_MODULE_DIR}/*")
+    list(APPEND CUDF_BASE_SRC ${CUDF_BASE_FILES})
+endforeach()
+
+add_library("${CUDF_NAMESPACE}_base" ${CUDF_BASE_SRC})
+add_library("${CUDF_NAMESPACE}::base" ALIAS "${CUDF_NAMESPACE}_base")
+add_dependencies("${CUDF_NAMESPACE}_base" stringify_run)
+
+foreach(CUDF_MODULE_DIR IN LISTS CUDF_MODULE_LIST)
+    string(  # replace "/" with ";" to create a list of directory leaves for the module directory
+        REPLACE "/" ";"
+            MODULE_DIR_LEAVES
+            ${CUDF_MODULE_DIR}
+    )
+    list(REVERSE MODULE_DIR_LEAVES)
+    list(GET MODULE_DIR_LEAVES 0 CUDF_MODULE_NAME)  # name the module after the last directory leaf
+    add_library_module(${CUDF_NAMESPACE} ${CUDF_MODULE_NAME} ${CUDF_MODULE_DIR})
+    target_link_libraries("${CUDF_NAMESPACE}_${CUDF_MODULE_NAME}" "${CUDF_NAMESPACE}::base")
+    list(APPEND CUDF_MODULES "${CUDF_NAMESPACE}_${CUDF_MODULE_NAME}")
+endforeach()
+
+foreach(CUDF_HEADERS_DIR IN LISTS CUDF_HEADERS_DIRS)
+    file(GLOB_RECURSE HEADERS "${CUDF_HEADERS_DIR}/*.hpp")
+    list(APPEND CUDF_HEADERS ${HEADERS})
+    file(GLOB_RECURSE HEADERS "${CUDF_HEADERS_DIR}/*.cuh")
+    list(APPEND CUDF_HEADERS ${HEADERS})
+endforeach()
+
+file(WRITE ${CUDF_SRC} "// This file is generated by ${CMAKE_CURRENT_LIST_FILE} \n")
+foreach(CUDF_HEADER IN LISTS CUDF_HEADERS)
+    file(APPEND ${CUDF_SRC} "#include \"${CUDF_HEADER}\" \n")
+endforeach()
+
+add_library(cudf ${CUDF_SRC})
+add_library(cudf::cudf ALIAS cudf)
+add_dependencies(cudf ${CUDF_MODULES})
+add_dependencies(cudf stringify_run)
 
 ###################################################################################################
 # - link libraries --------------------------------------------------------------------------------
 
-# link targets for cuDF
-target_link_libraries(cudf arrow arrow_cuda nvrtc ${CUDART_LIBRARY} cuda ${ZLIB_LIBRARIES} ${Boost_LIBRARIES})
+target_link_libraries(cudf_base arrow arrow_cuda nvrtc ${CUDART_LIBRARY} cuda ${ZLIB_LIBRARIES} ${Boost_LIBRARIES})
+target_link_libraries(cudf cudf_base ${CUDF_MODULES} arrow arrow_cuda nvrtc ${CUDART_LIBRARY} cuda ${ZLIB_LIBRARIES} ${Boost_LIBRARIES})
```

Comment:
Could we not relist all of the other libraries since they're linked to `cudf_base` already?

Reply:
Yes. This was a change I forgot to "undo" when I was trying out various things. It won't have an impact on the builds themselves because symbols won't be exported twice.